### PR TITLE
replaced Twitter futures with Scala futures

### PR DESCRIPTION
**Pull request by nsauro**
removed all Awaits except for one that is used by
MultiCollins/BackgroundProcess

This is one PR which removes Twitter futures and uses Scala Futures/timeouts.  I removed all timeouts that were not connected to multi collins(not used), and for background processing(because changing BackgroundProcess to being asynchronous is not a trivial task).

I have another branch off of this one which removes the finagle library entirely.  I will pr that one separately as I wanted to keep this one focused.

### CODE REVIEW

File: app/controllers/actors/AssetCancelProcessor.scala [hunk 1]
```diff
@@ -2,43 +2,51 @@ package controllers
...
-                  Left(Api.getErrorMessage("There was an error cancelling this server"))
-                case ticketId =>
-                  Asset.inTransaction {
-                    MetaWrapper.createMeta(asset, Map("CANCEL_TICKET" -> ticketId.toString))
-                    AssetLifecycle.updateAssetStatus(asset, AStatus.Cancelled, None, reason)
-                  }
-                  Await.result(plugin.setNote(n, "Cancelled: %s".format(reason)))
-                  Right(ticketId)
-              }
+        SoftLayer.pluginEnabled.fold(softlayerDisabled){ plugin =>
```
Comment (nsauro) [inline app/controllers/actors/AssetCancelProcessor.scala]: This change looks a lot bigger than it is.  I was having trouble following it, so I decided to move some of the constants into separate declarations.  I also opted to use the "fold" function instead of map/getOrElse.

File: app/util/concurrent/BackgroundProcessor.scala [hunk 2]
```diff
@@ -36,19 +35,28 @@ object BackgroundProcessor {
...
+    val mpd: Future[RESPONSE] = f.map{x => result((None, Some(x)))}
+    mpd.recover {
+      case t : TimeoutException => result((Some(SexyTimeoutException(cmd.timeout)), None))
+      case th : Throwable => result((Some(th), None))
+    }
+  }
+
+  //possibly the WORST name ever.  Don't judge me.  'flatSend' because the above function is really just a fancy map function
+  //since you pass a func that goes from x->y and get a future[y], this is just the flatMap version of that.
+  def flatSend[PROC_RES,RESPONSE](cmd: BackgroundProcess[PROC_RES])(result: SendType[PROC_RES] => Future[RESPONSE])(implicit mf: Manifest[PROC_RES]): Future[RESPONSE] = {
```
Comment (nsauro) [inline app/util/concurrent/BackgroundProcessor.scala]: Yep...worst name ever.  I needed a send function which allowed me to flatMap on the result.

File: app/collins/softlayer/SoftLayerPlugin.scala [hunk 3]
```diff
@@ -147,16 +149,27 @@ class SoftLayerPlugin(app: Application) extends Plugin with SoftLayer {
...
       true
-    } handle {
+    } recover {
       case e => false
     }
   }
 
   protected def makeRequest(request: HttpRequest): Future[HttpResponse] = {
     val client: Service[HttpRequest,HttpResponse] = clientSpec.build()
-    client(request) ensure {
```
Comment (maddalab) [inline app/collins/softlayer/SoftLayerPlugin.scala]: we should may be look into this in a different PR.

File: app/controllers/AssetWebApi.scala [hunk 4]
```diff
@@ -9,6 +9,7 @@ import util.concurrent.BackgroundProcessor
 
 import play.api.mvc._
 import play.api.libs.json._
+import play.api.libs.concurrent.Execution.Implicits._
```
Comment (maddalab) [inline app/controllers/AssetWebApi.scala]: Is this required?

File: app/controllers/actions/asset/ProvisionUtil.scala [hunk 5]
```diff
@@ -246,7 +246,7 @@ trait Provisions extends ProvisionUtil with AssetAction { self: SecureAction =>
     import play.api.Play.current
 
     val ActionDataHolder(asset, pRequest, _, attribs) = adh
-    BackgroundProcessor.send(ProvisionerTest(pRequest)(request)) { res =>
+    BackgroundProcessor.send(ProvisionerTest(pRequest)(request, defaultContext)) { res =>
```
Comment (maddalab) [inline app/controllers/actions/asset/ProvisionUtil.scala]: I see the implicit imported, is the specification of defaultContext required?

File: app/controllers/actors/ProvisionerProcessor.scala [hunk 6]
```diff
@@ -23,29 +22,33 @@ object ProvisionerStatus {
 
 case class ProvisionerResult(status: ProvisionerStatus, commandResult: CommandResult)
 
-case class ProvisionerTest(request: ProvisionerRequest, userTimeout: Option[FiniteDuration] = None)(implicit req: Request[AnyContent]) extends BackgroundProcess[ProvisionerResult]
+case class ProvisionerTest(request: ProvisionerRequest, userTimeout: Option[FiniteDuration] = None)(implicit req: Request[AnyContent], ec : ExecutionContext) extends BackgroundProcess[ProvisionerResult]
```
Comment (maddalab) [inline app/controllers/actors/ProvisionerProcessor.scala]: i think making ec implicit will avoid the specification of defaultContext aboe

File: app/collins/softlayer/SoftLayerPlugin.scala (hunk 3 above)
Comment (nsauro) [inline app/collins/softlayer/SoftLayerPlugin.scala]: The conversion of the the Twitter future to a Scala future?

File: app/controllers/AssetWebApi.scala (hunk 4 above)
Comment (nsauro) [inline app/controllers/AssetWebApi.scala]: AssetCancelProcessor needs an ExecutionContext since it is using Scala futures/Awaits to handle the Softlayer.cancelServer call.

File: app/controllers/actions/asset/ProvisionUtil.scala (hunk 5 above)
Comment (nsauro) [inline app/controllers/actions/asset/ProvisionUtil.scala]: Yes, only because the second parameter list for implicits is being manually invoked because 'request' is not an implicit value.

File: app/collins/softlayer/SoftLayerPlugin.scala (hunk 3 above)
Comment (nsauro) [inline app/collins/softlayer/SoftLayerPlugin.scala]: If its any help, I have another PR, right behind this which changes this class to not use the finagle library, and only to use the Play WS library.  That scraps all this conversion code, as the WS library returns Scala futures natively.

**Comment by byxorna** [comment]
This doesnt have any test coverage, right? I believe not, as @maddalab showed with his 2.1.5 PR. Do you think its worth it to figure out the test framework and get this covered, or is this a one-and-done thing?

File: app/controllers/actors/ProvisionerProcessor.scala (hunk 6 above)
Comment (nsauro) [inline app/controllers/actors/ProvisionerProcessor.scala]: ec is implicit here.  If you want to specify implicit parameters in a scala function, the parameter list needs to begin with keyword 'implicit', and then every parameter in that list is implicit.

http://stackoverflow.com/questions/10635406/defining-a-function-with-multiple-implicit-arguments-in-scala

**Comment by maddalab** [comment]
lgtm

**Comment by maddalab** [comment]
I tested this today, I am going to go ahead and merge this, feel free to grab me a nick at work if required.
